dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/utils/ProgramBuilder.java

```diff
@@ -78,6 +81,16 @@ public Program build() {
         return program;
     }
 
+    public void setZeroRegister(String registerName) {
+        for(Thread thread : program.getThreads()) {
+            for(Register register : thread.getRegisters()) {
+                if(register.getName().equals(registerName)) {
+                    register.setAsZeroRegister();
+                }
+            }
+        }
+    }
+
```

Comment:
Why are we back to such stateful approach?
Why not just provide the 0-register (its name) to the `ReplaceZeroRegisters` pass on construction and just invoke that pass inside the parser?
I think it is fair to say that this particular program transformation is something the parser or the program builder should do. You could even have the program builder take an optional set of passes as input that are run as post-processing (of the parsing) when building the final program. I would keep it simple and just invoke it directly from the parser though.

Reply:
After the bad experience we had with the boogie parser, I thought we decided against doing any kind of magic in the parser. The parser should parse, the transformation pipeline should transform.

Reply:
Function inlining etc. are a totally different thing: they are semantic-preserving transformations that we do either out of necessity or optimizations.
Zero registers, however, are a different thing because what the parser produces does simply not represent the correct semantics and we require an extra pass to "make it correct".
If zero registers were a feature of our core language, I wouldn't mind, but I think they are not and I don't see why they should be. I mean, in this PR you can mark a register as "zero" but it will not be enforced at all unless the register was specifically marked so during parsing time.